### PR TITLE
Update package to ESM

### PR DESCRIPTION
**Pull request by inferrinizzard**
- updated webpack and babel to ESM
- added additional entry points for each language
- TODO: debug tree shaking

### CODE REVIEW

**Comment by nene** [comment]
I'm pretty confident this setup won't work as indented.

In each of these files in `export/` dir you're importing `format` from `src/sqlFormatter.ts` which depends on all the languages being loaded.

File: .babelrc [hunk 1]
```diff
@@ -1,7 +1,18 @@
...
+        "modules": false
+      }
+    ],
+    "@babel/preset-typescript"
+  ],
   "plugins": [
-    "@babel/plugin-proposal-class-properties",
+    ["@babel/plugin-proposal-class-properties", { "loose": false }],
+    ["@babel/plugin-proposal-private-methods", { "loose": false }],
+    ["@babel/plugin-proposal-private-property-in-object", { "loose": false }],
```
Comment (nene) [inline .babelrc:15]: I don't think we need these. Especially as we use typescript which has its own syntax for private fields.

**Comment by nene** [comment]
For the tree-shaking to work we need to decouple language definitions from the `format()` function. For example as I roughly outlined in #452:

- We should export a base-formatter-function which takes `language: Formatter` option.
- And also export each of the `*Formatter` classes we have.

We probably don't need separate paths for the exports, something like this should work as well:

```ts
import { baseFormat, BigQueryFormatter } from 'sql-formatter`;

baseFormat("SELECT *", {language: BigQueryFormatter});
```

I didn't manage to come up with a good name for it though (`baseFormat` doesn't sound right). As it's really an alternative API for using the formatter, it might indeed be better to expose it under a separate path, like:

```ts
import { format, BigQueryFormatter } from 'sql-formatter/core`;
```

It might also be better to name these exported language definitions with simpler names, like:

```ts
import { format, bigquery } from 'sql-formatter/core`;
```

This way the name will be less implementation-dependant.

**Comment by nene** [comment]
There's currently one additional catch with making such a change. The `ExpressionFormatter.formatOperator()` contains currently a check `this.cfg.language === 'plsql'`. This won't work when language parameter is no more a string. We really should eliminate this language-specific switch from there and instead have some other config option like `denseAtOperator: boolean`. Currently there's no place for defining such language-specific formatting-configs. This could be achieved by adding a `formatterOptions()` method to `Formatter` class, which could be overridden by subclasses (in addition to the current `tokenizer()` method).

**Comment by nene** [comment]
In general I think it's better to split this whole task into two independent parts:

- providing an ESM version of sql-formatter.
- decoupling individual language definitions from the core formatter.

File: .babelrc [hunk 2]
```diff
@@ -1,7 +1,18 @@
...
+      {
+        "loose": true,
+        "modules": false
+      }
+    ],
+    "@babel/preset-typescript"
+  ],
   "plugins": [
-    "@babel/plugin-proposal-class-properties",
+    ["@babel/plugin-proposal-class-properties", { "loose": false }],
```
Comment (nene) [inline .babelrc:13]: `loose` option defaults to `false`, so it's unnecessary to specify it here.

File: .babelrc [hunk 3]
```diff
@@ -1,7 +1,18 @@
 {
-  "presets": ["@babel/preset-env", "@babel/preset-typescript"],
+  "presets": [
+    [
+      "@babel/preset-env",
+      {
+        "loose": true,
```
Comment (nene) [inline .babelrc:6]: Why turn on the loose mode?

File: package.json [hunk 4]
```diff
@@ -3,11 +3,17 @@
   "version": "10.7.1",
   "description": "Format whitespace in a SQL query to make it more readable",
   "license": "MIT",
-  "main": "lib/index.js",
+  "main": "dist/sql-formatter.min.js",
+  "module": "lib/index.js",
   "types": "lib/src/index.d.ts",
+  "exports": {
+    ".": "./lib/index.js"
+  },
```
Comment (nene) [inline package.json:11]: I wonder if it's really worth it to specify both `module` and `exports` fields. The `module` field was a proposed way of specifying exported module, but was never made official. Later the `exports` field was adopted in its place. Support for `exports` was added 2019 in NodeJS 12, which is no more actively supported.

I think `main` should serve as a good-enough fallback for those using some old NodeJS or some old bundler.

Comment (inferrinizzard) [reply package.json:11]: you're correct in that module is not an official spec for npm packaging purposes but it is used unofficially by webpack for their new experimental ESM export config

File: .babelrc (hunk 1 above)
Comment (inferrinizzard) [reply .babelrc:15]: it errors otherwise with the loose transform but I'll actually turn that off

**Review by nene — APPROVED** [review]
LGTM

**Comment by inferrinizzard** [comment]
Merging this as an initial conversion to ESM, tree-shaking to be adding in a later PR

**Pull request merged**